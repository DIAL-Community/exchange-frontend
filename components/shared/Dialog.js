import { useIntl } from 'react-intl'
import { FaSpinner } from 'react-icons/fa'
import { Dialog as DialogHeadlessui, Transition } from '@headlessui/react'
import { Fragment, useRef } from 'react'
import classNames from 'classnames'

export const DialogType = {
  FORM: 'form',
  DETAILS: 'details'
}

const Dialog = ({
  children,
  isSubmitInProgress,
  formId,
  isOpen,
  onClose,
  submitButton,
  cancelButton,
  closeButton,
  dialogType = DialogType.DETAILS
}) => {
  const { formatMessage } = useIntl()
  const format = (id, values) => formatMessage({ id }, values)

  const initialFocusRef = useRef(null)

  return (
    <Transition appear show={isOpen} as={Fragment} data-testid='dialog'>
      <DialogHeadlessui initialFocus={initialFocusRef} as='div' className='fixed z-100' onClose={onClose} >
        <div className="fixed inset-0 bg-dial-gray opacity-80" />
        <div className="fixed inset-0 flex items-center justify-center">
          <Transition.Child
            as={Fragment}
            enter='ease-out duration-300'
            enterFrom='opacity-0 scale-95'
            enterTo='opacity-100 scale-100'
            leave='ease-in duration-200'
            leaveFrom='opacity-100 scale-100'
            leaveTo='opacity-0 scale-95'
          >
            <DialogHeadlessui.Panel
              className={classNames({
                'max-h-screen w-screen md:w-3/5 lg:w-3/5 xl:w-1/2 p-8': dialogType === DialogType.FORM,
                'w-4/5 sm:w-4/5 md:w-3/5 lg:w-2/5 xl:w-2/6 p-2 sm:p-4': dialogType === DialogType.DETAILS
              }, 'transform bg-white shadow-xl rounded-2xl flex flex-col')}>
              <DialogHeadlessui.Description
                as='div'
                className={classNames({
                  'bg-edit': dialogType === DialogType.FORM,
                  'bg-white': dialogType === DialogType.DETAILS
                }, 'overflow-auto py-4 px-8 flex flex-col gap-3')}
              >
                <div className='flex justify-start text-xl' data-testid='dialog-body'>
                  {children}
                </div>
                <div className='flex justify-start gap-3 text-xl' >
                  {submitButton && (
                    <button
                      type='submit'
                      form={formId}
                      className='submit-button'
                      disabled={isSubmitInProgress}
                      data-testid='submit-button'
                    >
                      {format(`${isSubmitInProgress ? 'app.submitting' : 'app.submit'}`)}
                      {isSubmitInProgress && <FaSpinner className='spinner ml-3 inline' data-testid='submit-spinner' />}
                    </button>
                  )}
                  {cancelButton && (
                    <button
                      ref={initialFocusRef}
                      type='button'
                      onClick={onClose}
                      className='cancel-button'
                      disabled={isSubmitInProgress}
                      data-testid='cancel-button'
                    >
                      {format('app.cancel')}
                    </button>
                  )}
                </div>
                {closeButton && (
                  <div className='flex justify-end text-xl' >
                    <button
                      ref={initialFocusRef}
                      type='button'
                      onClick={onClose}
                      className='close-button'
                      disabled={isSubmitInProgress}
                      data-testid='close-button'
                    >
                      {format('app.close')}
                    </button>
                  </div>
                )}
              </DialogHeadlessui.Description>
            </DialogHeadlessui.Panel>
          </Transition.Child>
        </div>
      </DialogHeadlessui>
    </Transition>
  )
}

export default Dialog
