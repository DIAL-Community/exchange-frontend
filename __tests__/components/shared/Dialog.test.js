import { fireEvent } from '@testing-library/react'
import Dialog, { DialogType } from '../../../components/shared/Dialog'
import {
  mockObserverImplementation,
  mockRouterImplementation,
  render
} from '../../test-utils'
import CustomMockedProvider from '../../utils/CustomMockedProvider'

jest.mock('next/dist/client/router')

describe('Unit test for the Dialog component.', () => {
  const DIALOG_TEST_ID = 'dialog'
  const SUBMIT_BUTTON_TEST_ID = 'submit-button'
  const CANCEL_BUTTON_TEST_ID = 'cancel-button'
  const CLOSE_BUTTON_TEST_ID = 'close-button'
  const DIALOG_BODY_TEST_ID = 'dialog-body'
  const mockDialogBody = 'Mock Dialog Body'
  const mockOnClose = jest.fn()

  beforeAll(() => {
    mockRouterImplementation()
    window.IntersectionObserver = mockObserverImplementation()
  })


  test('Should not be visible when is not open.', () => {
    const { queryByTestId } = render(
      <CustomMockedProvider>
        <Dialog isOpen={false} />
      </CustomMockedProvider>
    )
    expect(queryByTestId(DIALOG_TEST_ID)).not.toBeInTheDocument()
  })

  describe('Should match snapshot -', () => {
    test('when Dialog has form type with submit and cancel button.', async () => {
      const { getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <Dialog
            submitButton
            cancelButton
            isOpen={true}
            onClose={mockOnClose}
            dialogType={DialogType.FORM}
          >
            {mockDialogBody}
          </Dialog>
        </CustomMockedProvider>
      )
      expect(getByTestId(DIALOG_BODY_TEST_ID)).toHaveTextContent(mockDialogBody)
      expect(getByTestId(SUBMIT_BUTTON_TEST_ID)).toHaveTextContent('Submit')
      expect(getByTestId(CANCEL_BUTTON_TEST_ID)).toHaveTextContent('Cancel')
      expect(queryByTestId(CLOSE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(getByTestId(DIALOG_TEST_ID)).toMatchSnapshot()
    })
      
    test('when Dialog has default type and has close button.', () => {
      const { getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <Dialog
            closeButton
            isOpen={true}
            onClose={mockOnClose}
          >
            {mockDialogBody}
          </Dialog>
        </CustomMockedProvider>
      )
      expect(getByTestId(DIALOG_BODY_TEST_ID)).toHaveTextContent(mockDialogBody)
      expect(getByTestId(CLOSE_BUTTON_TEST_ID)).toHaveTextContent('Close')
      expect(queryByTestId(SUBMIT_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(queryByTestId(CANCEL_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(getByTestId(DIALOG_TEST_ID)).toMatchSnapshot()
    })
  })

  describe('Should call', () => {
    test('the onClose function after clicking the "Cancel" button.', () => {
      const { getByTestId } = render(
        <CustomMockedProvider>
          <Dialog
            submitButton
            cancelButton
            isOpen={true}
            onClose={mockOnClose}
            dialogType={DialogType.FORM}
          >
            {mockDialogBody}
          </Dialog>
        </CustomMockedProvider>
      )
      fireEvent.click(getByTestId(CANCEL_BUTTON_TEST_ID))
      expect(mockOnClose).toBeCalled()
    })

    test('the onClose function after clicking the "Close" button.', () => {
      const { getByTestId } = render(
        <CustomMockedProvider>
          <Dialog
            closeButton
            isOpen={true}
            onClose={mockOnClose}
          >
            {mockDialogBody}
          </Dialog>
        </CustomMockedProvider>
      )
      fireEvent.click(getByTestId(CLOSE_BUTTON_TEST_ID))
      expect(mockOnClose).toBeCalled()
    })
  })
})
