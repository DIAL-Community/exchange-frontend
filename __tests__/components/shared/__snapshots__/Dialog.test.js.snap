// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the Dialog component. Should match snapshot - when Dialog has default type and has close button. 1`] = `
<div
  aria-describedby="headlessui-description-10"
  aria-modal="true"
  class="fixed z-100"
  data-testid="dialog"
  id="headlessui-dialog-7"
  role="dialog"
>
  <div
    class="fixed inset-0 bg-dial-gray opacity-80"
  />
  <div
    class="fixed inset-0 flex items-center justify-center"
  >
    <div
      class="w-4/5 sm:w-4/5 md:w-3/5 lg:w-2/5 xl:w-2/6 p-2 sm:p-4 transform bg-white shadow-xl rounded-2xl flex flex-col ease-out duration-300 opacity-0 scale-95"
      id="headlessui-dialog-panel-9"
    >
      <div
        class="bg-white overflow-auto py-4 px-8 flex flex-col gap-3"
        id="headlessui-description-10"
      >
        <div
          class="flex justify-start text-xl"
          data-testid="dialog-body"
        >
          Mock Dialog Body
        </div>
        <div
          class="flex justify-start gap-3 text-xl"
        />
        <div
          class="flex justify-end text-xl"
        >
          <button
            class="close-button"
            data-testid="close-button"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the Dialog component. Should match snapshot - when Dialog has form type with submit and cancel button. 1`] = `
<div
  aria-describedby="headlessui-description-5"
  aria-modal="true"
  class="fixed z-100"
  data-testid="dialog"
  id="headlessui-dialog-2"
  role="dialog"
>
  <div
    class="fixed inset-0 bg-dial-gray opacity-80"
  />
  <div
    class="fixed inset-0 flex items-center justify-center"
  >
    <div
      class="max-h-screen w-screen md:w-3/5 lg:w-3/5 xl:w-1/2 p-8 transform bg-white shadow-xl rounded-2xl flex flex-col ease-out duration-300 opacity-0 scale-95"
      id="headlessui-dialog-panel-4"
    >
      <div
        class="bg-edit overflow-auto py-4 px-8 flex flex-col gap-3"
        id="headlessui-description-5"
      >
        <div
          class="flex justify-start text-xl"
          data-testid="dialog-body"
        >
          Mock Dialog Body
        </div>
        <div
          class="flex justify-start gap-3 text-xl"
        >
          <button
            class="submit-button"
            data-testid="submit-button"
            type="submit"
          >
            Submit
          </button>
          <button
            class="cancel-button"
            data-testid="cancel-button"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
